### ISSUE TITLE
Widening/Narrowing conversion with lambda capture

### ISSUE THREAD

**Issue by dgrove-oss** [qbicc/qbicc#1356]
This test program (extracted from larger examples) results in generating invalid llvm IR.  The problem is related to lexical capture of a boolean.  The function that creates the lambda has an i32; the invocation of that function has an i1.

```
import java.lang.reflect.AccessibleObject;
import java.security.AccessController;
import java.security.PrivilegedAction;

public class LambdaCapture {
    public static void setAccessible(AccessibleObject o, boolean flag) {
        AccessController.doPrivileged((PrivilegedAction<Void>) () -> {
                try {
                    o.setAccessible(flag);
                } catch (SecurityException ex) {
                    ex.printStackTrace();
                }
                return null;
            });
    }

    public static void main(String[] args) {
        setAccessible(null, false);
    }
}
```

Error:
```
/tmp/output/LambdaCapture.ll: error: /Users/dgrove/bin/opt: /Users/dgrove/bin/opt: <stdin>:278:397: error: argument is not of expected type 'i32'
  %L2 = call i8 addrspace(1)* (i8 addrspace(1)*, i8 addrspace(1)*, i8 addrspace(1)*, i32) @"_Jjava_lang_invoke_LambdaForm$DMH$36_newInvokeSpecial__Ljava_lang_Object_2Ljava_lang_Object_2I_Ljava_lang_Object_2"(i8 addrspace(1)* %thr0, i8 addrspace(1)* addrspacecast (%"class.java.lang.invoke.DirectMethodHandle$Constructor"* @qbicc_initial_heap_obj_54494 to i8 addrspace(1)*), i8 addrspace(1)* %p0, i1 zeroext %L1) "statepoint-id"="94905", !dbg !788 ; LambdaCapture.java:7 bci@2
```

**Comment by dgrove-oss** [comment]
I also tried a variant when the captured variable was a `short` instead of a `boolean`.  Same problem.  So it's most likely a general missing widen operation at the call site , not something specific to booleans.

**Comment by dgrove-oss** [comment]
I think what's going on here is our ReflectionIntrinsics for linking the method handles need to be prepared to do the same demote/promote operations that we do in MethodParser for the invokebytecodes.

**Comment by dgrove-oss** [comment]
In particular, the LambdaForm generation code is being clever and trying to reduce the number of generated classes by collapsing boolean/byte/short/char/int into int.  So we need to accommodate that (the descriptor in the class file has the "real" type of the parameter, eg Z or S, but the Lambda constructor we get back is going to have I).

**Comment by dgrove-oss** [comment]
Fixed by #1359
